### ISSUE TITLE
Modifiche codice

### ISSUE THREAD

**Issue by Ale210501** [EsameUrban/Codice#1]
import pandas as pd
import numpy as np
from catboost import CatBoostClassifier
from sklearn.model_selection import StratifiedKFold, train_test_split
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
from collections import defaultdict
from tqdm import tqdm
import time

class GenomicCatBoostClassifier:
    def __init__(self, df, test_size=0.2, n_splits=5, random_state=42, balance_classes=True):
        """
        Initialize the CatBoost classifier for genomic data.
        """
        self.df = df.copy()
        self.test_size = test_size
        self.n_splits = n_splits
        self.random_state = random_state
        self.balance_classes = balance_classes
        self.model = None
        self.feature_importance = None
        self.cv_results = defaultdict(list)
        self.label_encoder = LabelEncoder()
        self.X_test = None
        self.y_test = None
        self.categorical_features = None

    def fix_categorical_features(self):
        """Fix categorical features by converting floats to strings."""
        print("🔧 Fixing categorical features...")

        potential_categorical = [
            col for col in self.df.columns
            if col not in ['patient_id', 'category'] and 
            (self.df[col].dtype == 'object' or 'miRNA_region' in col)
        ]
        print(f"   Found {len(potential_categorical)} potential categorical columns")

        fixed_count = 0
        with tqdm(potential_categorical, desc="Fixing categorical columns", 
                  bar_format='{l_bar}{bar:30}{r_bar}', colour='blue') as pbar:
            for col in pbar:
                original_dtype = self.df[col].dtype
                self.df[col] = self.df[col].astype(str).replace('nan', 'missing')
                if original_dtype != 'object':
                    fixed_count += 1
                    pbar.set_postfix({'Fixed': fixed_count})
                time.sleep(0.01)
        
        print(f"✅ Fixed {fixed_count} columns with float/numeric issues")
        return self

    def load_and_prepare_data(self):
        """Prepare data for training and testing."""
        print("📊 Preparing dataset...")

        self.fix_categorical_features()
        X = self.df.drop(['patient_id', 'category'], axis=1)
        y = self.label_encoder.fit_transform(self.df['category'])

        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=self.test_size, stratify=y, random_state=self.random_state
        )
        self.X_test, self.y_test = X_test, y_test

        categorical_features = [
            col for col in X.columns if X[col].dtype == 'object' or 'miRNA_region' in col
        ]
        self.categorical_features = categorical_features

        print(f"✅ Dataset prepared: {X.shape[0]} samples, {X.shape[1]} features")
        print(f"    Train: {X_train.shape[0]} samples, Test: {X_test.shape[0]} samples")
        print(f"    🏷️  Categorical features: {len(categorical_features)}")

        return X_train, y_train, categorical_features

    def train_with_stratified_cv(self):
        """Train the model using Stratified K-Fold CV on the training set and log metrics for each fold."""
        X_train, y_train, cat_features = self.load_and_prepare_data()

        skf = StratifiedKFold(n_splits=self.n_splits, shuffle=True, random_state=self.random_state)
        print(f"🎯 Starting Stratified {self.n_splits}-Fold CV on training set...")

        base_params = {
            'iterations': 1000,
            'learning_rate': 0.05,
            'depth': 6,
            'l2_leaf_reg': 3,
            'loss_function': 'Logloss',
            'eval_metric': 'AUC',
            'random_seed': self.random_state,
            'cat_features': cat_features,
            'verbose': False,
            'early_stopping_rounds': 50,
            'use_best_model': True
        }

        if self.balance_classes:
            base_params['auto_class_weights'] = 'Balanced'

        cv_splits = list(skf.split(X_train, y_train))
        all_fold_metrics = []

        with tqdm(total=self.n_splits, desc="Cross-Validation Progress", colour='cyan') as pbar:
            for fold, (tr_idx, val_idx) in enumerate(cv_splits, start=1):
                X_tr, X_val = X_train.iloc[tr_idx], X_train.iloc[val_idx]
                y_tr, y_val = y_train[tr_idx], y_train[val_idx]

                try:
                    model = CatBoostClassifier(**base_params)
                    model.fit(X_tr, y_tr, eval_set=(X_val, y_val), plot=False)

                    y_pred = model.predict(X_val)
                    y_proba = model.predict_proba(X_val)[:, 1]
                    auc = roc_auc_score(y_val, y_proba)
                    prec = precision_score(y_val, y_pred, average='weighted')
                    rec = recall_score(y_val, y_pred, average='weighted')
                    f1 = f1_score(y_val, y_pred, average='weighted')

                    self.cv_results['auc'].append(auc)
                    self.cv_results['precision'].append(prec)
                    self.cv_results['recall'].append(rec)
                    self.cv_results['f1'].append(f1)
                    self.cv_results['feature_importances'].append(model.get_feature_importance())

                    fold_metrics = {
                        'Fold': fold,
                        'AUC': auc,
                        'Precision': prec,
                        'Recall': rec,
                        'F1': f1
                    }
                    all_fold_metrics.append(fold_metrics)
                    print(f"Metrics for Fold {fold}: {fold_metrics}")

                except Exception as e:
                    print(f"Error during Fold {fold}: {str(e)}")
                    raise e

                pbar.update(1)

        if self.cv_results['feature_importances']:
            self.feature_importance = np.mean(self.cv_results['feature_importances'], axis=0)

        print("\n📘 Training final model on full training set...")
        final_params = base_params.copy()
        self.model = CatBoostClassifier(**final_params)
        self.model.fit(X_train, y_train)

        return all_fold_metrics

    def print_cv_summary(self):
        """Print cross-validation results summary."""
        if not self.cv_results['auc']:
            print("No CV results available")
            return
            
        print("\n📈 Cross-Validation Summary:")
        print("=" * 50)
        
        metrics = ['auc', 'precision', 'recall', 'f1']
        for metric in metrics:
            values = self.cv_results[metric]
            if values:
                mean_val = np.mean(values)
                std_val = np.std(values)
                print(f"{metric.upper():>10}: {mean_val:.4f} (±{std_val:.4f})")
        
        print("=" * 50)

    def evaluate_on_test(self):
        """Evaluate the trained model on the hold-out test set."""
        if self.model is None or self.X_test is None:
            raise ValueError("Model not trained or test split not available.")

        print("\n🧪 Evaluating on hold-out test set...")
        y_true = self.y_test
        y_pred = self.model.predict(self.X_test)
        y_proba = self.model.predict_proba(self.X_test)[:, 1]

        report = classification_report(y_true, y_pred, target_names=self.label_encoder.classes_)
        auc = roc_auc_score(y_true, y_proba)
        cm = confusion_matrix(y_true, y_pred)

        print("\n📋 Classification Report:")
        print(report)
        print(f"\n🎯 ROC AUC: {auc:.4f}")
        print("\n📊 Confusion Matrix:")
        print(cm)

        return {
            'classification_report': classification_report(y_true, y_pred, output_dict=True),
            'roc_auc': auc,
            'confusion_matrix': cm
        }

    def get_top_features(self, top_n=20):
        """Return top N most important features."""
        if self.feature_importance is None:
            print("❌ Feature importance not available - train model first")
            return None
            
        features = self.df.drop(['patient_id', 'category'], axis=1).columns
        imp_df = pd.DataFrame({
            'feature': features,
            'importance': self.feature_importance
        }).sort_values('importance', ascending=False)
        
        print(f"\n🏆 Top {top_n} Most Important Features:")
        print("=" * 60)
        print(imp_df.head(top_n))
        print("=" * 60)
        
        return imp_df.head(top_n)
